docs/synapse/userguides/storm_adv_functions.rstorm

```diff
@@ -0,0 +1,777 @@
+     pipeline). In this case the nodes themselves don't "cause" the function to execute; the pipeline runs
+     once per node, and the function invoked by the pipeline runs once each time the pipeline runs.
+
+- Nodes do not "pass into" functions by default, as the function and the invoking Storm logic are two separate
+  pipelines. It is possible to **invoke** a function so that it operates on a node or nodes; but the function
+  will not "automatically" do so.
+  
+The sections below on :ref:`storm-func-nodes` and :ref:`storm-func-runtsafe` discuss these behaviors in more detail.
+
+.. _storm-func-basics:
+
+Function Basics
+---------------
+
+Storm supports three types of functions. Each is explained in more detail below.
+
+- **callable** functions, which are "regular" functions (similar to functions in other programming languages).
+  Callable functions return a value.
+- **data emitter** functions, which emit data.
+- **node yielder** functions, which yield nodes.
+
+.. _storm-func-declare:
+
+Declaring Functions
++++++++++++++++++++
+
+All functions are **declared** in the same way, using the ``function`` keyword.
+
+- ``function myFunction() { <do stuff> }``
+- ``function myFunction(foo) { <do stuff> }``
+- ``function myFunction(bar,baz=$lib.null) { <do stuff> }``
+
+.. _storm-func-invoke:
+
+Invoking Functions
+++++++++++++++++++
+
+All functions are **invoked** using the function name preceded by the dollar sign ( ``$`` ), and by
+passing any required parameters to the function.
+
+- ``$myFunction()``
+- ``$myFunction(foo)``
+- ``$myFunction($foo)``
+- ``$myFunction(bar,baz=wheee)``
+
+Parameters can be passed as literals, variables, or keyword=$valu arguments. For example, given a
+function that takes an organization name as input:
+
+::
+  
+  function $myFunction(orgname) { <do stuff> }
+
+The name can be passed directly:
+
+::
+  
+  $myFunction("The Vertex Project")
+
+...as a variable:
+
+::
+  
+  $name="The Vertex Project"
+  $myFunction($name)
+
+...or as a keyword/value pair:
+
+::
+  
+  $myFunction(orgname="The Vertex Project")
+
+::
+  
+  $name="The Vertex Project"
+  $myFunction(orgname=$name)
+
+
+.. _storm-func-nodes:
+
+Operating on Nodes
+******************
+
+Functions do not inherit or operate on the invoking Storm pipeline by default. If you want a function
+to operate on nodes in a pipeline, you must invoke the function in such a way as to pass the node (or a
+property or properties of the node) as input to the function.
+
+For example, if your invoking pipeline consists of a set of ``inet:ipv4`` nodes, a function can take the
+``:asn`` property as input:
+
+::
+  
+  $myFunction(:asn)
+
+Or:
+
+::
+  
+  $asn=:asn
+  $myFunction($asn)
+
+Alternatively, you can pass the entire ``$node`` object to the function and use the ``yield`` keyword
+within the function to yield the node into the function's pipeline:
+
+::
+  
+  //Declare function
+  function $myFunction(node) {
```

Comment:
Lets remove the use of the variable `$node` from the documentation. We typically use the variable `$n` or something descriptive like `$inboundNode` to avoid any confusion with the runtime populated `$node` variable.

Reply:
revised